frame/data-feed/src/lib.rs

```diff
@@ -0,0 +1,832 @@
+pub struct FeededDataInfo<BlockNumber> {
+	/// the key of what we want from the json callback
+	key_str: Vec<u8>,
+	number_type: NumberType,
+	operation: Operations,
+	schedule: BlockNumber,
+}
+
+impl<BlockNumber> FeededDataInfo<BlockNumber> {
+	fn key_str(&self) -> &[u8] {
+		&self.key_str
+	}
+
+	fn number_type(&self) -> NumberType {
+		self.number_type
+	}
+}
+
+pub trait Trait: CreateSignedTransaction<Call<Self>> {
+	type Event: From<Event<Self>> + Into<<Self as frame_system::Trait>::Event>;
+	/// The identifier type for an offchain worker.
+	type AuthorityId: AppCrypto<Self::Public, Self::Signature>;
+	/// The origin that can control `StorageKey`, control `ActiveProviders` and other key information
+	type DataFeedOrigin: EnsureOrigin<Self::Origin>;
+	/// WeightInfo
+	type WeightInfo: WeightInfo;
+}
+
+decl_event!(
+	pub enum Event<T>
+	where
+		AccountId = <T as frame_system::Trait>::AccountId,
+		BlockNumber = <T as frame_system::Trait>::BlockNumber,
+	{
+		/// Register a storage key
+		RegisterStorageKey(StorageKey),
+		/// Remove a storage key
+		RemoveStorageKey(StorageKey),
+		/// Add a provider accountid
+		ProviderAdded(AccountId),
+		/// Remove a provider accountid
+		RemoveProvider(AccountId),
+		/// Set url for a key, which would be used in offchain
+		SetUrl(StorageKey, Vec<u8>),
+		/// New data stored on blockchain
+		FeedData(AccountId, StorageKey, DataType),
+		/// Set calculated data
+		SetData(StorageKey, Option<DataType>),
+		/// Set offchain
+		SetOffchainPeriod(StorageKey, Option<BlockNumber>),
+	}
+);
+
+// Errors inform users that something went wrong.
+decl_error! {
+	pub enum Error for Module<T: Trait> {
+		/// If the storage key already registered
+		ExistingKey,
+		/// You can not put data under the key that is not registered
+		InvalidKey,
+		/// If the data type does not match the one specified in DataInfo
+		InvalidValue,
+		/// User is not allowed to feed data onto the blockchain
+		NotAllowed,
+		/// Provider count of key count exceed limit
+		ExceededLimit,
+		/// Not allow to use offchain to submit.
+		NotAllowOffchain,
+	}
+}
+
+decl_storage! {
+	trait Store for Module<T: Trait> as DataFeed {
+		/// A set of storage keys
+		/// A storage key in this set refers to a storage value in other pallets,
+		pub ActiveParamTypes get(fn all_keys): Vec<StorageKey>;
+
+		/// A set of accounts
+		/// Any one can submit data on chain, while only the data submitted by
+		/// the provider in this set is valid, and will be calculated later.
+		pub ActiveProviders get(fn all_providers): Vec<T::AccountId>;
+
+		/// Data Attributes.
+		/// It defines the type of the data, how to extract the data we want
+		/// from a json blob
+		pub DataInfo get(fn data_info): map hasher(twox_64_concat) StorageKey => Option<FeededDataInfo<T::BlockNumber>>;
+
+		/// Permissible URL that could be used to fetch data
+		pub Url get(fn url): map hasher(twox_64_concat) StorageKey => Option<Vec<u8>>;
```

Comment:
This is more of a meta comment - I think it makes more sense to have individual data providers be able to set their own URL, with a defined default. Otherwise the actual usefulness of the output is questionable.

Reply:
It can't be done here for Offchain worker do not provide a way to get AccountId ( current only provide all/any method to do something, but there is no way to get specific accountid, even can't get all accountid in keystore), if offchain worker provide this feature, this could do.

Reply:
https://github.com/paritytech/substrate/pull/6960/files/1f97c8be51eb988b77be731a3598615385b6ba5e#diff-6b937194476b7433d97c87af8802bbe4R741
I mark this as todo